### PR TITLE
[SYCL] [NATIVECPU] Support barriers on Native CPU

### PR DESCRIPTION
**Pull request by PietroGhg**
This PR adds support to barriers by integrating LLVM passes from the [oneAPI Construction Kit](https://github.com/codeplaysoftware/oneapi-construction-kit). The OCK is fetched via CMake when the Native CPU plug-in is enabled, and can optionally be disabled through the `NATIVE_CPU_USE_OCK` CMake option. When the OCK is available, the `NATIVECPU_USE_OCK` definition is set for the targets defined in `llvm/lib/SYCLLowerIR`.

This PR also refactors the pass pipeline for Native CPU, making it so all of passes relevant to Native CPU are run in the `clang` invocation that lowers the IR to the target ISA, this is done to allow correct handling of barriers with functions declared with `SYCL_EXTERNAL`.

Unified Runtime PR: https://github.com/oneapi-src/unified-runtime/pull/1152

### CODE REVIEW

File: clang/test/CodeGenSYCL/native_cpu_basic.cpp
Comment (premanandrao) [inline clang/test/CodeGenSYCL/native_cpu_basic.cpp:52]: Does this test check name mangling specific to native cpu anymore?   If not, can we fix the comments here, at the top and the typeinfo name check below?

File: llvm/include/llvm/SYCLLowerIR/SYCLNativeCPUPipeline.h
Comment (bader) [inline llvm/include/llvm/SYCLLowerIR/SYCLNativeCPUPipeline.h:1]: License header must be added to all new source files!
Please, add here and to llvm/lib/SYCLLowerIR/PipelineSYCLNativeCPU.cpp.

File: clang/test/CodeGenSYCL/native_cpu_basic.cpp
Comment (PietroGhg) [reply clang/test/CodeGenSYCL/native_cpu_basic.cpp]: Yes you are right, it's not specific to native cpu anymore so I've removed the check, thank you

File: llvm/include/llvm/SYCLLowerIR/SYCLNativeCPUPipeline.h
Comment (PietroGhg) [reply llvm/include/llvm/SYCLLowerIR/SYCLNativeCPUPipeline.h]: Sorry, I've added the licence headers

**Review by premanandrao — APPROVED** [review]
CFE part LGTM.

**Review by mdtoguchi — APPROVED** [review]
OK for driver

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:209]: [message()](https://cmake.org/cmake/help/latest/command/message.html) command expects a _mode_ as the first argument. In this case I believe `WARNING` would be appropriate.

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:202]: Seems odd to have `NATIVECPU_USE_OCK` and `Native_CPU_USE_OCK`. What's the difference between them?

The documentation has a different spelling again `NATIVE_CPU_USE_OCK`.

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (PietroGhg) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: Yes you are right, I've added `STATUS` and `WARNING` where appropriate, thank you

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (PietroGhg) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: This was a mistake on my part, thanks for spotting it, it should be fixed now

**Review by kbenzie — APPROVED** [review]
UR LGTM

**Comment by wenju-he** [comment]
@PietroGhg there is merge conflict

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp:309]: Do I understand correctly that we now emit those functions as `external` ones? Should we maybe make them `internal` instead to enable more aggressive optimizations with them? If those functions are external, they may as well be defined in a library written in C++ instead of being on-demand generated at LLVM IR level

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp:434]: Should we grab at least some of debug information from the original instruction instead of just resetting it completely?

**Review by AlexeySachkov — APPROVED** [review]
Changes in a few files I'm a code owner of looks good to me

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (PietroGhg) [reply llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp]: Yes you are right, I've changed that code to only set an empty debug location if no previous location existed, thank you

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (PietroGhg) [reply llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp]: Yes, thanks for spotting this

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (PietroGhg) [reply llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp:309]: You are right, setting `internal` makes most sense here, I've done it

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (PietroGhg) [reply llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp:395]: I've added a comment explaining why we need this, thank you

File: sycl/doc/design/SYCLNativeCPU.md
Comment (PietroGhg) [reply sycl/doc/design/SYCLNativeCPU.md:74]: At this stage we'd like to keep the option of at least being able to compile trivial applications without the OCK, but yeah it the medium term it would be good to drop the mode without the OCK since it would simplify testing and maintenance, @uwedolinsky what do you think?

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (AlexeySachkov) [reply sycl/plugins/unified_runtime/CMakeLists.txt:59]: If that is the case I would maybe suggest to move the PR to draft state to prevent accidental merge once the last code owner approves

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp:309]: Note that no changes were done to this file, I still see the same removed line without any replacement - a missed `git commit` perhaps?

File: sycl/doc/design/SYCLNativeCPU.md
Comment (uwedolinsky) [reply sycl/doc/design/SYCLNativeCPU.md:74]: I agree. We only test with OCK enabled going forward so it makes sense dropping the mode without OCK.

File: llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp
Comment (PietroGhg) [reply llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp:309]: I've made the change at the end of `addReplaceFunc` [here](https://github.com/intel/llvm/blob/faa3be6ba1f969c583091d5bb2d7c5267ac4fb74/llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp#L307) so that it applies to all the functions that we insert in the module

**Review by sarnex — APPROVED** [review]
llvm/lib/SYCLLowerIR/CMakeLists.txt LGTM, sorry for late review

**Comment by PietroGhg** [comment]
I've marked this PR as Draft until https://github.com/oneapi-src/unified-runtime/pull/1152 is merged and the UR tag is updated

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (kbenzie) [inline sycl/plugins/unified_runtime/CMakeLists.txt:66]: https://github.com/oneapi-src/unified-runtime/pull/1152 has now merged so please pull in the latest sycl branch changes, resolve the conflicts, update the UR tag, then make this ready for review. 

```suggestion
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit eddda12352c27961fb454331f88e05d5335db7ed
  # Merge: 6892412f 71e7e38e
  # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
  # Date:   Wed Jan 24 12:59:32 2024 +0000
  #     Merge pull request #1152 from PietroGhg/pietro/barriers
  #     [NATIVECPU] Handle oneAPI Construction Kit integration in runtime
  set(UNIFIED_RUNTIME_TAG eddda12352c27961fb454331f88e05d5335db7ed)
```

**Review by kbenzie — APPROVED** [review]
UR LGTM

**Comment by PietroGhg** [comment]
Hello @intel/llvm-gatekeepers, this is ready to be merged, thank you

**Pull request merged**
